### ISSUE TITLE
zipkin2 with cassandra3 consume a lot of disk space

### ISSUE THREAD

**Issue by dumbdonkey** [openzipkin/zipkin#1861]
I have upgrade to zipkin 2.3 version. well, I have found the annotation_query_index use so many disk space which would be 50 times bigger than data file.

```

-rw-r--r-- 1 root root  5638620424 Dec 21 23:57 mc-209-big-SI_span_annotation_query_idx.db
-rw-r--r-- 1 root root     5898292 Dec 21 23:57 mc-209-big-SI_span_l_service_idx.db
-rw-r--r-- 1 root root     1566280 Dec 21 23:57 mc-214-big-Index.db
-rw-r--r-- 1 root root       83192 Dec 21 23:57 mc-214-big-Filter.db
-rw-r--r-- 1 root root       14596 Dec 21 23:57 mc-214-big-Summary.db
-rw-r--r-- 1 root root          10 Dec 21 23:57 mc-214-big-Digest.crc32
-rw-r--r-- 1 root root   106792591 Dec 21 23:57 mc-214-big-Data.db

```

I have google some docs about this problem. blog <http://www.doanduyhai.com/blog/?p=2058> give a detailed explanation and it suggest 

```
As a conclusion, use CONTAINS mode wisely and be ready to pay the price in term of disk space. There is no way to avoid it. Even with efficient search engines like ElasticSearch or Solr, it is officially recommended to avoid substring search (LIKE %substring%) for the sake of performance.

```

is there any good suggestion about this situation ?

**Comment by codefromthecrypt** [comment]
First question is if you use zipkin's search by key/value or not? This is only used for that feature

It is possible we could allow a config to whitelist keys to index (like http.path)

@michaelsembwever any thoughts?

**Comment by codefromthecrypt** [comment]
by the way, thank you very much for the detailed info!

**Comment by dumbdonkey** [comment]
@adriancole I have add a whitelist by myself. which still consume lot of space. I plan to develop a customed TokenizingAnalyzer.

**Comment by michaelsembwever** [comment]
> It is possible we could allow a config to whitelist keys to index (like http.path)

@adriancole, a whitelist sounds like a good idea to start with.

beyond that: the only reason we're using `CONTAINS` here is because SASI on collections types are not supported. see https://issues.apache.org/jira/browse/CASSANDRA-11182

i'm thinking the right way to move forward would be to denormalise the annotations like…
```
CREATE TABLE IF NOT EXISTS zipkin2.span_by_annotation_query (
    trace_id            text, // when strictTraceId=false, only contains right-most 16 chars
    ts_uuid             timeuuid,
    id                  text,
    trace_id_high       text, // when strictTraceId=false, contains left-most 16 chars if present
    annotation_query    text, 
    PRIMARY KEY         (trace_id, ts_uuid, id, annotation_query)
);

CREATE CUSTOM INDEX IF NOT EXISTS ON zipkin2.span_by_annotation_query (annotation_query) USING 'org.apache.cassandra.index.sasi.SASIIndex'
   WITH OPTIONS = {'mode': 'PREFIX'};
```
Each annotation would be written to `annotation_query`, instead of the `░` joined string.

And the look-up would then mean in `SelectTraceIdsFromSpan`…

 - the statement would instead be prepared with `QueryBuilder.like("annotation_query", bindMarker("annotation_query"))`
 - and the statement called with  just `annotationKey` instead of `"%░" + annotationKey + "░%"`.

This would allow removing the `span.annotation_query` column and its `CONTAINS` SASI.

A custom TokenizingAnalyzer is another approach, and would be very cool to see.

**Comment by codefromthecrypt** [comment]
@zuochangan did you end up doing  a TokenizingAnalyzer? how did it go?

**Comment by dumbdonkey** [comment]
@adriancole I have finished it.    But I'm a little busy now. After the completion of the work at hand, I will share with solutions.

**Comment by codefromthecrypt** [comment]
Thank you for following up!

**Comment by dumbdonkey** [comment]
Sorry for late reply. 

Here is my solution. 

In our situation, we won't query with partial annotation , we always query with a complete annotation. so prefix index and proper TokenAnalyzer is enough. Here are steps you can follow. 

1. compile following code and package it into a jar then put it in cassandra library path and restart cassandra.

```


package org.apache.cassandra.index.sasi.analyzer;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;
import java.util.Locale;
import java.util.Map;

import com.google.common.base.Strings;

import org.apache.cassandra.db.marshal.AbstractType;

public class ZipkinStandardAnalyzer extends AbstractAnalyzer
{
    private static final Locale DEFAULT_LOCALE = Locale.getDefault();
    private static final String ZIPKIN_DELIMITER = "░";
    private AbstractType validator;
    private List<String> tokens = new ArrayList<>();
    private Iterator<String> iter;
    
    public ZipkinStandardAnalyzer()
    {
    }

    public ByteBuffer next()
    {
        return this.validator.fromString(iter.next());
    }

    public void init(Map<String, String> options, AbstractType validator)
    {
        this.validator = validator;
    }

    public boolean hasNext()
    {
        return iter.hasNext();
    }

    public void reset(ByteBuffer input)
    {
        this.next = null;
        this.tokens.clear();
        String ann = validator.getString(input);
        ann = ann.toLowerCase(DEFAULT_LOCALE);
        String[] parts = ann.split(ZIPKIN_DELIMITER);
        for (String part : parts)
        {
            if (!Strings.isNullOrEmpty(part))
            {
                tokens.add(part);
            }
        }
        this.iter = tokens.iterator();
    }


    public boolean isTokenizing()
    {
        return true;
    }
}


```


2. replace old annotation index with following statement


```
CREATE CUSTOM INDEX span_annotation_query_idx ON zipkin2.span (annotation_query) USING 'org.apache.cassandra.index.sasi.SASIIndex' WITH OPTIONS = {'analyzer_class': 'org.apache.cassandra.index.sasi.analyzer.ZipkinStandardAnalyzer', 'case_sensitive': 'false', 'mode': 'prefix', 'analyzed': 'true'};

```

3. modify class zipkin2.storage.cassandra.SelectTraceIdsFromSpan as following and deploy your zipkin with new code again.

```

Call<Set<Entry<String, Long>>> newCall(
      @Nullable String serviceName,
      String annotationKey,
      TimestampRange timestampRange,
      int limit
    ) {
      Input input = new AutoValue_SelectTraceIdsFromSpan_Input(
        serviceName,
        // Notice here,we just need to remove all ░
         annotationKey,
        timestampRange.startUUID,
        timestampRange.endUUID,
        limit
      );
      return new SelectTraceIdsFromSpan(
        this,
        serviceName != null ? withServiceAndAnnotationQuery : withAnnotationQuery,
        input
      ).flatMap(new AccumulateTraceIdTsLong());
    }


```



That's all. Hope I've made myself clear.

**Comment by codefromthecrypt** [comment]
This is cool. can you mention any results (ex difference in space used or
indexing or query latency?

**Comment by dumbdonkey** [comment]
@adriancole following is current disk space usage

Index Size/ Data Size ratio is 0.15. I remeber old ratio is 60, even bigger. 

![image](https://user-images.githubusercontent.com/16414369/34904936-5eefbcb0-f88a-11e7-86ec-b8f2c01a041e.png)

Since we save a lot of spans every day. query latency is very high with annotation_query index.  most of the time, we just got a timeout error. I have move column 'annotation_query' to table 'trace_by_service_span' . that gives us a reasonable response time.

ex:

an query like 
http://*/zipkin/?serviceName=all&spanName=all&startTs=1515834112059&endTs=1515837712059&minDuration=&limit=10&annotationQuery=error&sortOrder=duration-desc

query with annotation_index on table span will show timeout error at last.

query with annotation_index on table trace_by_service_span has a latency lower than 5 seconds, which is acceptable.

![image](https://user-images.githubusercontent.com/16414369/34905037-c8ab9484-f88b-11e7-9c8f-cc5e1707cde9.png)

**Comment by llinder** [comment]
Storage has been a pain point for me as well and finally to the point that I need to consider options.  Both of the suggestions here would work for me.  Using a different TokenAnalyzer is a really nice idea and I'm open to adopting that approach.  My only concern is that it might not be something new users can easily adopt and might be a source of confusion long term.

Short term the TokenAnalyzer would be an easier "quick fix" and at the same time still allow for adopting a longer term solution later without data loss.

If I was to invest some resource to assist on an improvement, which of these two would others like to see as a long term solution?

**Comment by michaelsembwever** [comment]
@llinder ,
 you can now just drop the `CONTAINS SASI`, a la the annotation query index.

If you don't need search by annotation in the UI this is by far the simplest solution.

this was done in https://github.com/openzipkin/zipkin/pull/1902

**Comment by michaelsembwever** [comment]
@zuochangan, your token analyzer would be a good patch for Cassandra.

If the delimiter character was parameterised, then as a `DelimiterTokenAnalyzer` i think it would be a patch that would get excepted. And we could then make the use of it permanent in zipkin.

**Comment by codefromthecrypt** [comment]
great idea.. you mean good patch upstream in cassandra? If so, I'm happy to
help review.

**Comment by michaelsembwever** [comment]
> If the delimiter character was parameterised, then as a DelimiterTokenAnalyzer i think it would be a patch that would get excepted. And we could then make the use of it permanent in zipkin.

https://issues.apache.org/jira/browse/CASSANDRA-14247

Feel free to offer a review @adriancole and @zuochangan 
Also the nomenclature, which i'm kinda struggling with.

Any testing on this would do wonders, if you found time, @zuochangan .

**Comment by llinder** [comment]
@michaelsembwever search in the UI is super important to us and one of the drivers for our early adoption of the Cassandra3 span store vs the original Cassandra span store.

Sounds like things might be leaning towards a TokenAnalyzer approach.  For now I will try and adopt that approach as well and contribute, review or test where I can.

**Comment by michaelsembwever** [comment]
@zuochangan, your work has been committed to Cassandra. It will become available in the next release: Cassandra-3.11.3

And an added win: an interesting suggestion by @mkjellman saw a ~60% performance improvement at write time.

**Comment by tramchamploo** [comment]
Seems the `DelimiterAnalyzer.java` works out fine.
And I find this jaeger's solution here. [Schema](https://github.com/jaegertracing/jaeger/blob/master/plugin/storage/cassandra/schema/v001.cql.tmpl)
Using another table for tag index: 
```
CREATE TABLE IF NOT EXISTS ${keyspace}.tag_index (
    service_name    text,
    tag_key         text,
    tag_value       text,
    start_time      bigint,
    trace_id        blob,
    span_id         bigint,
    PRIMARY KEY ((service_name, tag_key, tag_value), start_time, trace_id, span_id)
)
```
Can anyone tell the  difference of performance between these two?

**Comment by michaelsembwever** [comment]
@tramchamploo,
> `PRIMARY KEY ((service_name, tag_key, tag_value) …`

You're going to get some very hot partition keys there. And there's no way of de-duping it because you want each span_id in the results.
This is only going to work if you have reverse clustering on start_time, TWCS, TTL, and queries are time-bound.
That said it's a better denormalisation approach over what i suggested above in https://github.com/openzipkin/zipkin/issues/1861#issuecomment-353510778, which in hindsight was a bit of a brain-fart.

**Comment by alseddnm** [comment]
@adriancole @michaelsembwever I believe this what we need for #2094 .. similar problem I think.. we are getting lots write timeout
<img width="469" alt="screen shot 2018-06-19 at 4 29 13 pm" src="https://user-images.githubusercontent.com/4561575/41682520-91de6cae-749d-11e8-9343-31754a2094f1.png">

we did disable the search feature because of this,  we need to visualize our data in the UI.
I already started looking into Elastic or Kafka datastore , I will try this fix 1st.

**Comment by codefromthecrypt** [comment]
missed this note, but quick clarification which is indexing by (tag, value,
service) -> traceid should be ok even without spanid. query is still valid
as it is based on service. this lowers writes significantly with deduping
for apps that make many local spans.

this isnt to counter any other points made.

**Comment by codefromthecrypt** [comment]
ignore me.. I also had a brain fart! yeah span id is needed when combining
conditions.

**Comment by codefromthecrypt** [comment]
fixed in zipkin 2.11
